modules/adwords_appengine/src/main/java/com/google/api/ads/adwords/jaxws/utils/v201802/SelectorBuilderImpl.java

```diff
@@ -0,0 +1,472 @@
+   */
+  private Set<Predicate> copyPredicatesSet() {
+    Set<Predicate> predicatesCopy = Sets.newLinkedHashSet();
+    for (Predicate predicate : this.predicates) {
+      Predicate copyPredicate = new Predicate();
+      copyPredicate.setField(predicate.getField());
+      copyPredicate.setOperator(predicate.getOperator());
+      copyPredicate.getValues().addAll(predicate.getValues());
+      predicatesCopy.add(copyPredicate);
+    }
+    return predicatesCopy;
+  }
+
+  /**
+   * Copies the whole set of OrderBy objects, by creating new instances of each one in the given
+   * set.
+   *
+   * @return the new set of OrderBy objects with different instance for each of the OrderBy
+   */
+  private Set<OrderBy> copyOrderingSet() {
+    Set<OrderBy> orderingCopy = Sets.newLinkedHashSet();
+    for (OrderBy order : this.ordering) {
+      OrderBy copyOrder = new OrderBy();
+      copyOrder.setField(order.getField());
+      copyOrder.setSortOrder(order.getSortOrder());
+      orderingCopy.add(copyOrder);
+    }
+    return orderingCopy;
+  }
+
+  @Override
+  public SelectorBuilderImpl fields(String... fields) {
+    this.fields = Sets.newLinkedHashSet();
+    Collections.addAll(this.fields, fields);
+    return this;
+  }
+
+  @Override
+  public SelectorBuilderImpl fields(EntityField... fields) {
+    return fields(this.enumFieldsToNames(fields));
+  }
+
+  /**
+   * Returns the array of Strings resulting from calling {@code name()} on each field.
+   *
+   * @param fields the array of fields as Java Enums.
+   */
+  private String[] enumFieldsToNames(EntityField... fields) {
+    String[] fieldsNames = new String[fields.length];
+    for (int i = 0; i < fields.length; i++) {
+      fieldsNames[i] = fields[i].name();
+    }
+    return fieldsNames;
+  }
+
+  @Override
+  public SelectorBuilderImpl increaseOffsetBy(int additionalOffset) {
+    if (this.paging == null) {
+      this.paging = new Paging();
+    }
+    Integer startIndex = this.paging.getStartIndex();
+    if (startIndex == null) {
+      startIndex = 0;
+    }
+    this.paging.setStartIndex(startIndex + additionalOffset);
+    return this;
+  }
+
+  @Override
+  public SelectorBuilderImpl limit(int limit) {
+    if (this.paging == null) {
+      this.paging = new Paging();
+    }
+    this.paging.setNumberResults(limit);
+    return this;
+  }
+
+  @Override
+  public SelectorBuilderImpl offset(int offset) {
+    if (this.paging == null) {
+      this.paging = new Paging();
+    }
+    this.paging.setStartIndex(offset);
+    return this;
+  }
+
+  @Override
+  public SelectorBuilderImpl removeLimitAndOffset() {
+    this.paging = null;
+    return this;
+  }
+
+  @Override
+  public SelectorBuilderImpl orderAscBy(String field) {
+    OrderByWrapper order = new OrderByWrapper();
+    order.setField(field);
+    order.setSortOrder(SortOrder.ASCENDING);
+
+    this.ordering.add(order);
+
+    return this;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 184:190, 200:206.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.